src/validation.h

```diff
@@ -744,6 +745,62 @@ bool InvalidateBlock(CValidationState& state, const CChainParams& chainparams, C
 /** Remove invalidity status from a block and its descendants. */
 void ResetBlockFailureFlags(CBlockIndex* pindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
+/**
+ * Serialized version of a UTXO set from which an assumed-valid CChainState
+ * can be constructed.
+ */
+class SnapshotMetadata
+{
+public:
+    //! The hash of the block that reflects the tip of the chain for the
+    //! UTXO set contained in this snapshot.
+    uint256 m_base_blockhash;
+
+    //! A hash of the snapshot contents as generated by `GetUTXOStats`.
+    uint256 m_utxo_contents_hash = {};
+
+    //! The number of coins in the UTXO set contained in this snapshot.
+    uint64_t m_coins_count = 0;
+
+    //! Necessary to "fake" the base nChainTx so that we can estimate progress during
+    //! snapshot IBD.
+    unsigned int m_nchaintx = 0;
+
+    //! True if this snapshot has been validated by IBDing the chain from scratch up
+    //! to the base block of the snapshot (`m_base_blockhash`).
+    //!
+    //! This is serialized (vs. being memory only) because once we complete
+    //! validation of a snapshot, we update the metadata on disk so that the next
+    //! time bitcoind starts up, we know we've validated the snapshot. This would
+    //! be unnecessary if we overwrote the `chainstate` leveldb directory with the
```

Comment:
Don't get this field, it is serialized down to a `utxo.dat` file, which is going to be pass to other nodes, what's the meaning of validation completed for them ?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/16899#discussion_r326899181

In commit "validation: add unused SnapshotMetadata class" (3511d6b2ea65219a18dbb6d6ca45a4de62569dc1)

> Don't get this field, it is serialized down to a `utxo.dat` file, which is going to be pass to other nodes, what's the meaning of validation completed for them ?

I guess SnapshotMetadata is being used for two different things:

- As a header for dumptxoutset data
- As a way to track state if a node is shut down in the middle of background sync.

I think it would probably be clearer to use separate structs, maybe having the first one included in the second one to avoid duplication.

Reply:
Good points. I've removed this field for now. I'll defer figuring out exactly how to track this state to when I file the PR for loading snapshots, but I suspect I'll just use the presence of a `/datadir/chainstate_*` directory to indicate an unvalidated assumeutxo chainstate, then overwrite `/datadir/chainstate` on validation.